### DISCUSSION TITLE
Command palette (or: why static commands would be awesome to implement)

### DISCUSSION THREAD

**Discussion by Wilker-uwu** [VoylinsGamedevJourney/gozen#8] [Ideas]
hewwo!

so basically i love command palettes. it's always refreshing to see such a feature in any complex program, because i know i can always rely on it as a fallback to learning and even discovering anything in a program before i have the need to stop what i'm doing and look for tutorials.

![A screenshot showing-off the command palette in Lapce](https://github.com/voylin/GoZen/assets/10854855/855f416d-f256-44cf-8c7a-3500045b38a6 "A screenshot showing-off the command palette in Lapce.")
###### More about Lapce: https://lapce.dev/  ;  Command palette is briefly mentioned here: https://docs.lapce.dev/get-started/command-palette

but more important than that, is also knowing how powerful it can be for [making a program scalable using static functions](https://github.com/helix-editor/helix/blob/9690e0b733fd82e0ced65388a24ad80f769ce3b4/helix-term/src/commands.rs#L192), for both implementing, and making any user choices a lot more flexible, which i think it's gonna be pretty useful for the project  :3

![A demo gif showing off the usage of the Helix editor](https://github.com/voylin/GoZen/assets/10854855/8cab401a-ba29-4ef8-8347-4d99dc89b15c "A demo gif showing off the usage of the Helix editor, starting with browsing the tutor with a couple written commands, and ending with a brief showcase of my own settings for it.")
###### More about Helix text editor: https://helix-editor.com/  ;  User remapping is documented here: https://docs.helix-editor.com/remapping.html

in practice, an implementation would look like this:

- reserve a simple, easily reachable keybind (e.g `:`);
- such a keybind should open a single-line text box with a reserved character in the beginning -- when in doubt, choose a colon `:` like in most editors and CLIs, a greater-than `>` symbol like in VSCode, or just a slash `/` symbol like a lot of people who are familiar with Minecraft commands.
- deleting such a symbol should either cancel the prompt, or provide a project-wide file browsing list with all possible files;
- all searches should be done with a fuzzy search function, ranking results by closest first match, but not necessarily discarding any options not matching it;
- you might need to look for a solution for disambiguation early on. there's a chance that a plugin might look to replace a function or give the same name with a different meaning. i don't know any standards about this;
- subcommands are also a thing, both for keybinds and commands. the docs for the Helix editor has a more intuitive way of handling it in text, i think;
- bonus points if you decide to make full descriptions on top of keybind suggestions;

additionally, i think this should also make it possible to apply the same static function solution when implementing keybinds to the program. if you need a sequence of keys, set up a "leading key", or multiple ones depending on the configuration and then its subcommands. for instance, `Ctrl+q` and then `!` would quit the program quit without saving as if typing in `:quit!` or `:q!` to quit as well (exclamation mark means ignore failsafes, like checking if the project had any unsaved changes), but `Ctrl+q` and `s` would `:write-quit` or `:wq`, and `Ctrl+q` and then **`S`** (notice the uppercase letter) would be the equivalent of "Save" and "Save As...".

i think this should make the project much easier going forward.

what do you think? :3

**Comment by voylin** [comment]
A command palette has been one of the things I've been considering for GoZen. I want this editor to look and feel minimalistic (not too much clutter going on), so using something like this would probably be the best way to go. The implementation is something I've been wondering about as I want all parts of the editor to be modular, so this interface would have to be an autoload. 

I could see myself focusing on this too much and hiding buttons inside the main editor screen, expecting people to use the command palette but I don't think this would be beneficial for some people so I may add an option in the settings for enabling 'Zen mode' which basically eliminated most buttons which aren't used very frequently allowing for more space for the timeline and project view I guess.

I'm terrible at designing UX, but I'm learning so let's see which designs I can come up with to implement this. But first things first, the focus right now should be on making the editor use-able, I will start working on designing this as I feel that it would be greatly beneficial to the project.

Thanks for your time to suggest this and give lots of detailed information and examples! Really helpful!
